### ISSUE TITLE
tile.transform

### ISSUE THREAD

**Issue by curran** [d3/d3-tile#4]
Currently, setting the scale and transform based on a zoom transform is a bit cumbersome:

``` js
var transform = d3.event.transform;
var tiles = tile
  .scale(transform.k)
  .translate([transform.x, transform.y])
  ();
```

It would be nice to be able to say this instead:

``` js
var transform = d3.event.transform;  
var tiles = tile.transform(transform)();
```

Here's a [prototype implementation for this](http://bl.ocks.org/curran/defa669c2e7935a91ef5457fe5dd3bc2) where the setter looks like this:

``` js
tile.transform = function(_) {
  if(arguments.length){
    return tile.scale(_.k).translate([_.x, _.y]);
  } else {
    return d3.zoomIdentity.translate(tx, ty).scale(scale);
  }
};
```

**Comment by mbostock** [comment]
Something that irks me is that *tile* doesn’t take any arguments, so you end up with these trailing parentheses whenever you use it:

```js
var tiles = d3.tile()
    .size([width, height])
    .scale(transform.k)
    .translate([transform.x, transform.y])
    ();
```

The verbosity of setting *tile*.scale and *tile*.translate from a zoom transform doesn’t bother me especially, though, since if you are using a projection, you end up having to do the same thing with the projection—albeit slightly customized to the projection’s scale definition. For Mercator:

```js
projection
    .scale(transform.k / tau)
    .translate([transform.x, transform.y]);
```

So, if you end up fixing one but not the other, it’s a little awkward.

It also bothers me that *transform*.translate takes two arguments (for symmetry with *context*.translate on a Canvas 2D context) whereas *tile*.translate and *projection*.translate takes one (a two-element array; so that they can be used as both getters and setters).

A thought I had was that *tile*.scale and *tile*.translate could be defined as accessors based on the input you pass to *tile*…

```js
var tile = d3.tile()
    .size([width, height])
    .scale(function(t) { return t.k; })
    .translate(function(t) { return [t.x, t.y]; });
```

Then you could pass the *transform* to *tile* and it would extract the necessary parts:

```js
var tiles = tile(transform);
```

And, the default *tile*.scale and *tile*.translate could be defined as above to be compatible with zoom transforms.

I suppose that’s an improvement, but it still leads to some asymmetry with projections:

```js
var tiles = tile(transform);
projection.scale(transform.k / tau).translate([transform.x, transform.y]);
```

I’m not sure how to avoid that. But, it does make me feel like changing *tile*.scale and *tile*.translate to be accessors is overkill.

I suppose another option would be to have *tile* always take a transform and not have *tile*.scale and *tile*.translate accessors. And maybe you could also have a new *projection*.transform method which is a convenience method for setting *projection*.scale and *projection*.translate? Then something like this would work:

```js
var tile = d3.tile().size([width, height]);
var tiles = tile(transform);
projection.transform(transform.scale(1 / tau));
```

That seems pretty clean, I guess? Then there would be a soft dependency on d3-zoom but presumably you could also pass an object `{k, x, y}` in place of a bona fide zoom transform to *tile* and it would still work…

**Comment by mbostock** [comment]
Putting this under consideration for 1.0.

**Comment by mbostock** [comment]
It’s a little awkward right now that the *tile*.scale “bakes-in” the tile size. This is why *tile*.scale defaults to 256 instead of 1, and it’s why you have to multiply *transform*.k by the tile size to compute the desired *tile*.scale (from [Zoomable Map Tiles](https://observablehq.com/@d3/zoomable-map-tiles)):

```js
let tiles = tile
    .scale(transform.k * 256)
    .translate([transform.x, transform.y])
  ();
```

**Comment by mbostock** [comment]
(You can avoid multiplying the *transform*.k each time by changing the *zoom*.scale to be premultiplied by the tile size, but that feels a little weird.)

**Comment by mbostock** [comment]
Implemented scale and translate accessors in 45f949385b422214538a730a4bfcf7449b0919a3, so this works out of the box with d3-zoom’s transforms.
